ESPS/general/src/fftinv/fftinv.1
.\" Copyright (c) 1986-1990 Entropic Speech, Inc.
.\" Copyright (c) 1991 Entropic Research Laboratory, Inc. All rights reserved.
.\" @(#)fftinv.1	1.10 04 Jan 1993 ESI/ERL
.ds ]W (c) 1991 Entropic Research Laboratory, Inc.
.TH FFTINV 1\-ESPS 04 Jan 1993
.SH "NAME"

fftinv \- Inverse Fast Fourier Transform of ESPS FEA_SPEC file to FEA_SD file
.SH "SYNOPSIS"
.B fftinv
[
.BI \-r " range"
] [
.BI \-t " output_type"
] [
.BI \-x " debug_level"
]
.I specfile
.I sdfile
.SH "DESCRIPTION"
.PP
.I Fftinv
takes an input ESPS spectral record file (FEA_SPEC file), 
.I specfile,
and takes the inverse Fast Fourier Transform of one or more records
to produce an ESPS sampled-data file (FEA_SD file)
.I sdfile.
If the input file name
.I specfile
is replaced by "\-", then stdin is read; similarily, if
.I sdfile
is replaced by "\-", then the output is written to stdout.
.PP
The input FEA_SPEC file must have spectral type SPTYP_CPLX and 
frequency format SPFMT_ASYM_EDGE \- i.e., the full complex 
spectrum is stored.  Such spectral records are produced 
by \fIfft\fP (1\-\s-1ESPS\s+1) using the \fB\-c\fP option.  
.SH "OPTIONS"
.PP
The following options are supported:
.TP
.BI \-r " first" : "last"
.TP
.BI \-r " first" \- "last"
.TP
.BI \-r " first" :+ "incr"
In the first two forms, a pair of unsigned integers
specifies the range of input records to process.  If 
.IR last " = " first " + " incr,
the third form (with the plus sign) specifies the same range as the
first two forms.  If 
.I first
is omitted, the default value of 1 is used.  If 
.I last
is omitted, then all records from 
.I first
are processed.
.TP
.BI \-t " output_type"
Specifies the ESPS data type of the output sampled-data file.
Allowable values are DOUBLE, FLOAT, LONG, SHORT, BYTE,
DOUBLE_CPLX, FLOAT_CPLX, LONG_CPLX, SHORT_CPLX, and BYTE_CPLX.
Results are computed as float or float_cplx
and converted to the specified type.
If one of the 5 real types is specified, the output is the real part
of the inverse-transformed data; the imaginary part is simply discarded
with no check to determine whether it is negligible.
If the option is not specified, the generic header item
.I "input_data"
is read from the input file.
If if the item exists and has the string "real" as its value,
the output FEA_SD file is of type FLOAT
and contains only the real part of the inverse transform.
Otherwise the output FEA_SD file has type FLOAT_CPLX.
.TP
.BI \-x " debug_level\fR [0]\fP"
Specifies that debugging output be printed on stderr; larger values
of 
.I debug_level
result in more output.  The default is for no output.  
.SH "ESPS PARAMETERS"
.PP
The ESPS parameter file is not read.
.SH "ESPS COMMON"
.PP
The ESPS Common file is not read 
or written.
.SH "ESPS HEADERS"
.PP
The generic header items
.I "spec_type"
and
.I "freq_format"
are read from the input file
and checked to assure that they have the required values,
SPTYP_CPLX and SPFMT_ASYM_EDGE.
.PP
The generic header item
.I "num_freqs"
is read from the input file
and determines the length of the inverse Fourier transform;
the transform length is 1 less than the value of
.I "num_freqs"
because of the way spectral values are ordered in the ASYM_EDGE format: 
the zero-frequency value is in the middle,
preceded by the negative-frequency values
and followed by the positive-frequency values,
with the value at the folding frequency duplicated
as both the first and the last value.
The transform length is written as the header item
.I "inv_fft_length"
in the output file.
The generic header item
.I "fft_length,"
if it exists, is read from the input file and checked for consistency;
if it is different from the transform length determined from
.I "num_freqs,"
a warning is written on
.I stderr.
.PP
The generic header item
.I "sf"
is read from the input file if present
and determines the value used for the sampling frequency;
otherwise a default of 8000 Hz is used.
This value is written as the header item
.I "record_freq"
in the output file
and is used for scaling and other purposes during the computation.
.PP
The number of samples in the first input frame is obtained,
either from the generic header item
.I "frmlen"
or from the field
.I "frame_len"
in the first record.
One half this frame length is converted to seconds,
with the sampling frequency as the conversion factor;
the resulting value is the offset
from the beginning of the first input frame to the middle.
The
.I "start_time"
header item is read from the input file if present;
otherwise a default of 0 is used.
This is assumed to refer to the middle of the first frame.
The offset determined from the length of the first frame is subtracted
to get the time corresponding to the beginning of the first frame.
A second adjustment to this value determines the value to be written as the
.I "start_time"
header item of the output file:
adding the offset from the first record of the input file
to the first record actually processed.
.PP
If the
.B \-t
option is not specified, the generic header item
.I "input_data,"
if it exists, is read from the input file
and determines the data type of the output file as specified in the
discussion of the option.
.SH "FUTURE CHANGES"
.PP
.SH "SEE ALSO"
.PP
fft (1\-ESPS), plotsd (1\-ESPS), get_fft_inv (3\-ESPS), FEA_SD(5\-ESPS),
FEA_SPEC(5\-ESPS), ESPS(5\-ESPS).
.SH "BUGS"
.PP
None known.
.SH "REFERENCES"
.PP
J. W. Cooley and J. W. Tukey, "An Algorithm for the Machine Calculation of
Complex Fourier Series," \fIMath. Computation\fP, Vol. 19, 1965, pp. 297-301.
.sp
Alan V. Oppenheim and Ronald W. Schafer, \fIDigital Signal Processing\fP.
Englewood Cliffs, New Jersey: Prentice-Hall, Inc., 1975.
.SH "AUTHOR"
.PP
manual page and program by John Shore, using library
function by Shankar Narayan.
Modified to write FEA_SD files by David Burton.



ESPS/general/src/fftinv/fftinv.c
/*
 * This material contains proprietary software of Entropic Speech, Inc.
 * Any reproduction, distribution, or publication without the prior
 * written permission of Entropic Speech, Inc. is strictly prohibited.
 * Any public distribution of copies of this work authorized in writing by
 * Entropic Speech, Inc. must bear the notice
 *
 *   "Copyright (c) 1987 Entropic Speech, Inc.; All rights reserved"
 *
 * Program: fftinv.c	
 *
 * Written by:  John Shore
 * Checked by:
 *
 * This program computes inverse FFTs from FEA_SPEC files to SD files
 */

#ifndef lint
static char *sccs_id = "@(#)fftinv.c	1.13	8/31/95 ESI";  
#endif
char *Version = "1.13";
char *Date = "8/31/95";
char *ProgName = "fftinv";

#include <stdio.h>
#include <esps/esps.h>
#include <esps/unix.h>
#include <esps/feasd.h>
#include <esps/feaspec.h>
#include <esps/fea.h>

#define ERROR_EXIT(text) {(void) fprintf(stderr, "%s: %s - exiting\n", \
		ProgName, text); exit(1);}

#define REQUIRE(test,text) {if (!(test)) {(void) fprintf(stderr, \
"%s: %s - exiting\n", ProgName, text); exit(1);}}

#define SYNTAX \
 USAGE("fftinv [-r range] [-t output_type] [-x debug_level] specfile sdfile")

/*
 Globals
*/
int	    debug_level = 0;		/* debug level */

/*
 * system functions
 */
double  log(), sqrt();
/*
 * ESPS library functions
 */
void	set_sd_type();
char	*eopen();
char	*get_cmd_line();
void	lrange_switch(), get_fft_inv();
void    update_waves_gen();
struct feasd  *allo_feasd_recs();
int	is_type_complex();
/*
 * local functions
 */
static int is_type_numeric();

/*
 * main program
 */

main(argc, argv)
    int  argc;
    char **argv;
{
    FILE    *specstrm = stdin,	/* input and output file streams */
            *sdstrm = stdout;
    struct header  		/* input and output file headers */
            *ih, *oh;
    char    *specfile, 		/* input and output file names */
	    *sdfile;
	    
    struct feaspec  *spec_rec;   /* record for spectral data */
    struct feasd    *sd_rec;     /* record for feasd data */

    extern int
	    optind;
    extern char
            *optarg;
    int	    c;
    long    srec, erec;		/* start and end records */
    double  log2 = log(2.0);	/* needed for base 2 logs */
    char    *range = NULL;	/* string for range specification (-r) */
    int	    r_flag = 0;		/* flag for range option */
    float   *x, *y;		/* arrays for get_fftinv */
    float   scale;		/* scale factor from continuous to discrete */
    long    nan, hf, j;
    long    length;		/* transform length */
    int	    order;

    float   *fdata;		/* pointer to real data in feasd record */
    float_cplx
	    *cdata;		/* pointer to complex data in feasd record */
    long    *frame_len;		/* pointer to frame length */
    double  record_freq;	/* record_freq header item in input */
    double  sf;			/* sample frequency in original sampled data */
    double  start_time;         /* start time of beginning of first frame */
    int	    t_flag = NO;	/* -t option specified? */
    char    *output_type;	/* name of output data type */
    int	    type = FLOAT_CPLX;	/* type code for output data type */
    int	    real = NO;          /* YES for real inverse Xform; NO for complex */

    /*
     * process command line options
     */
    while ((c = getopt(argc, argv, "r:t:x:")) != EOF)
        switch (c)    
        {
	case 'x':
	    debug_level = atoi(optarg);
	    break;
        case 'r': 
            range = optarg;
	    r_flag++;
            break;
	case 't':
	    output_type = optarg;
	    t_flag++;
	    break;
        default: 
            SYNTAX;
            break;
        }

    srec = 1;
    erec = LONG_MAX;
    if (r_flag) 
	lrange_switch (range, &srec, &erec, 1);	

    if (srec < 1)
	ERROR_EXIT("can't have negative starting point");
    if (srec > erec)
	ERROR_EXIT("range must specify at least one record");
    nan = erec - srec + 1;
    if (debug_level) 
	Fprintf(stderr, "fftinv: range is %ld - %ld\n", srec, erec);

    /*
     * process file arguments - first, input file
     */

    if (optind < argc)
	specfile = 
	   eopen("fftinv", argv[optind++], "r", FT_FEA, FEA_SPEC, &ih, &specstrm);
    else {
	Fprintf(stderr, "fft: no input FEA_SPEC file specified.\n");
	SYNTAX;
    }    
    if (debug_level)
	Fprintf(stderr, "Input file is %s\n", specfile);

    if (t_flag)
    {
	type = lin_search(type_codes, output_type);
	if (!is_type_numeric(type))
	{
	    Fprintf(stderr,
	    	    "fftinv: invalid output type %s\n", output_type);
	    exit(1);
	}
    }
    else if (genhd_type("input_data", (int *) NULL, ih) == CHAR
	     && strcmp(get_genhd_c("input_data", ih), "real") == 0)
	type = FLOAT;
    else
	type = FLOAT_CPLX;

    /*
     * allocate spectral record
     */

    spec_rec = allo_feaspec_rec(ih, FLOAT); 

    /*
     * set up output file
     */

    if (optind < argc)
	sdfile = 
	    eopen("fftinv", argv[optind++], "w", NONE, NONE, &oh, &sdstrm);
    else {
	Fprintf(stderr, "fftinv: no output SD file specified.\n");
	SYNTAX;
    }
    if (debug_level)
	Fprintf(stderr, "Output file is %s\n", sdfile);

    REQUIRE(genhd_type("spec_type", (int *) NULL, ih) == CODED,
           "Header item \"spec_type\" undefined or wrong type");
    if (*get_genhd_s("spec_type", ih) != SPTYP_CPLX) 
	ERROR_EXIT("FEA_SPEC file spec_type is not complex (SPTYP_CPLX)");

    REQUIRE(genhd_type("freq_format", (int *) NULL, ih) == CODED,
           "Header item \"spec_type\" undefined or wrong type");
    if (*get_genhd_s("freq_format", ih) != SPFMT_ASYM_EDGE) 
	ERROR_EXIT("FEA_SPEC file freq_format is not SPTYP_ASYM_EDGE");

    length = *get_genhd_l("num_freqs", ih) - 1;
    hf = length / 2;
    order = ROUND(log((double) length) / log2);
    if (debug_level) 
	Fprintf(stderr, 
	    "fftinv: transform length = %ld, order = %ld, num_freqs = %ld\n", 
	    length, order, *get_genhd_l("num_freqs", ih));

    if (genhd_type("fft_length", (int *) NULL, ih) != HD_UNDEF)
    {
	if (get_genhd_val("fft_length", ih, -1.0) != (double) length)
	    Fprintf(stderr,
		"%s: fft_length + 1 not equal to num_freqs.\n",
		ProgName);
    }

    x = (float *) calloc((unsigned) length, sizeof(float));
    y = (float *) calloc((unsigned) length, sizeof(float));
    spsassert(x != NULL && y != NULL, 
	"fftinv: couldn't allocate enough memory");

    /*
     * create and fill in FEA_SD file header
     */

    oh = new_header(FT_FEA);
    add_source_file(oh, specfile, ih);
    add_comment(oh, get_cmd_line(argc, argv));
    oh->common.tag = NO;
    oh->variable.refer = ih->variable.refer;
    (void) strcpy(oh->common.prog, ProgName);
    (void) strcpy(oh->common.vers, Version);
    (void) strcpy(oh->common.progdate, Date);
    *add_genhd_l("inv_fft_length", (long *) NULL, 1, oh) = length;

    /*
     * compute frequency and start_time generics
     */

    sf = get_genhd_val("sf", ih, 8000.0);

    REQUIRE(genhd_type("frame_meth", (int *) NULL, ih) == CODED,
	    "Header item \"frame_meth\" undefined or wrong type");
    switch (*get_genhd_s("frame_meth", ih))
    {
    case SPFRM_FIXED:
    case SPFRM_VARIABLE:
	frame_len = spec_rec->frame_len;
	break;
    case SPFRM_NONE:
	/* Frame length not available.
	   Use transform length for lack of anything better. */
	frame_len = &length;
	break;
    default:
	REQUIRE(NO, "Unrecognized value of \"frame_meth\" in input header");
	break;
    }

    /* read record for length of first frame */

    REQUIRE(get_feaspec_rec(spec_rec, ih, specstrm) != EOF,
	    "Can't get first record of input file");

    start_time = 
	get_genhd_val("start_time", ih, 0.0)
	- (double) (*frame_len)/(2*sf);

    record_freq = get_genhd_val("record_freq", ih, 0.0);
    if (record_freq != 0.0)
	start_time += (srec - 1)/record_freq;
    
    /* move to starting position */

    if (srec > 1)
    {
	if (srec > 2)
	    fea_skiprec(specstrm, srec - 2, ih);
	REQUIRE(get_feaspec_rec(spec_rec, ih, specstrm) != EOF,
		"Can't get first input record of specified range");
    }
    nan--;

    /*
     * initialize FEA_SD header, allocate data structures,
     * and write output file header
     */

    real = !is_type_complex(type);

    REQUIRE(init_feasd_hd(oh, type, 1, &start_time, NO, sf) == 0,
	    "Couldn't initialize output FEA_SD header");
    sd_rec = allo_feasd_recs(oh, type, length, (char *) NULL, NO);
    REQUIRE(sd_rec != NULL, "Couldn't allocate FEA_SD data record");

    if (type == FLOAT_CPLX)
	cdata = (float_cplx *) sd_rec->data;
    else if (!real)
	cdata = (float_cplx *) malloc(length * sizeof(float_cplx));
	
    write_header(oh, sdstrm);

    /*
     * main loop
     */

    do
    {
	long	frmlen = *frame_len;

	if (debug_level) Fprintf(stderr, "fftinv: frame %ld\n", srec++);
	if (debug_level > 1) {
	    Fprintf(stderr, 
		"fftinv: real and imag components in spectral record:\n");
	    for (j = 0; j < length + 1; j++) 
		Fprintf(stderr, "%f\t%f\n", 
		    spec_rec->re_spec_val[j], spec_rec->im_spec_val[j]);
	}	
	/*
	 * compute factor to get from continuous density to discrete
	 * spectral distribution
	 */

	scale = sqrt(frmlen * sf);

	/*
	 * reorder spectral components to get_fftinv order
	 * remember that ASYM_EDGE stores frequencies (-hf, ..., 0, ..., hf),
	 * hf = length / 2
	 */
	for (j = 0; j < length; j++) {
	    if (j < hf) {
		    x[j] = scale * spec_rec->re_spec_val[j + hf];
		    y[j] = scale * spec_rec->im_spec_val[j + hf];
		}
	    else {
		    x[j] = scale * spec_rec->re_spec_val[j - hf];
		    y[j] = scale * spec_rec->im_spec_val[j - hf];
	    }
	}
	if (debug_level > 1) {
	    Fprintf(stderr, 
		"fftinv: real and imag spectral components input to get_fftinv:\n");
	    for (j = 0; j < length; j++) 
		Fprintf(stderr, "%f\t%f\n", x[j], y[j]);
	}	

	/* do inverse fft */
	get_fft_inv(x, y, order);

	if (debug_level > 1) {
	    Fprintf(stderr, "fftinv: sampled data output from get_fft_inv:\n");
	    for (j = 0; j < length; j++) 
		Fprintf(stderr, "%f\n", x[j]);
	}

	/* pack the data into the output record */

	if (real)
	    type_convert(frmlen, (char *) x, FLOAT,
			 (char *) sd_rec->data, type, (void (*)()) NULL);
	else
	{
	    for(j = 0; j < frmlen; j++)
	    {
		cdata[j].real = x[j];	  
		cdata[j].imag = y[j];
	    }
	    if (type != FLOAT_CPLX)
		type_convert(frmlen, (char *) cdata, FLOAT_CPLX,
			     (char *) sd_rec->data, type, (void (*)()) NULL);
	}

	REQUIRE((put_feasd_recs(sd_rec, 0L, frmlen, oh, sdstrm) == 0),
	 "Couldn't write output record");

    } while (nan-- && get_feaspec_rec(spec_rec, ih, specstrm) != EOF);

    exit(0);
    /*NOTREACHED*/
}


static int
is_type_numeric(type)
    int	    type;
{
    switch (type)
    {
    case BYTE:
    case SHORT:
    case LONG:
    case FLOAT:
    case DOUBLE:
    case BYTE_CPLX:
    case SHORT_CPLX:
    case LONG_CPLX:
    case FLOAT_CPLX:
    case DOUBLE_CPLX:
        return YES;
    default:
        return NO;
    }
}
